crates/miden-lib/asm/kernels/transaction/lib/faucet.masm

```diff
@@ -201,31 +197,65 @@ proc.burn_non_fungible_asset
 
     # prepare stack for removal of non-fungible asset from tracking SMT
-    swapw padw
-    # => [EMPTY_WORD, ASSET, SMT_ROOT, ASSET]
+    swapw exec.asset_vault::build_non_fungible_asset_vault_key padw swapw movup.12
+    # => [faucet_storage_data_slot, ASSET_KEY, EMPTY_WORD, SMT_ROOT, ASSET]
 
     # remove the non-fungible asset from the tracking SMT
-    exec.smt::set
-    # => [OLD_VAL, SMT_ROOT', ASSET]
-
-    # assert the `OLD_VAL` is not EMPTY_WORD, indicating that the non-fungible asset exists.
-    # we only need to check ASSET[1] as this is always set to the faucet_id and can not be 0.
-    drop drop eq.0 not assert.err=ERR_FAUCET_NON_FUNGIBLE_ASSET_TO_BURN_NOT_FOUND drop
-    # => [SMT_ROOT', ASSET]
+    exec.account::set_map_item dropw
+    # => [OLD_VAL, ASSET]
 
-    # update the root of the SMT containing the non-fungible assets
-    exec.account::get_faucet_storage_data_slot exec.account::set_item dropw
+    # Assert the `OLD_VAL` is not an EMPTY_WORD, indicating that the non-fungible asset exists. We 
+    # only need to check OLD_VAL[3] as this is always set to the faucet_id_prefix and can not be 0.
+    eq.0 not assert.err=ERR_FAUCET_NON_FUNGIBLE_ASSET_TO_BURN_NOT_FOUND drop drop drop
     # => [ASSET]
 
     # remove the non-fungible asset from the input vault for asset preservation checks
     exec.memory::get_input_vault_root_ptr movdn.4 exec.asset_vault::remove_non_fungible_asset
     # => [ASSET]
 end
 
+#! Returns a boolean indicating whether the provided non-fungible asset has been already issued by 
+#! this faucet.
+#!
+#! Inputs:  [ASSET]
+#! Outputs: [is_issued]
+#!
+#! Where:
+#! - ASSET is the non-fungible asset that is being checked.
+#! - is_issued is a boolean indicating whether the non-fungible asset has been issued.
+#!
+#! Panics if:
+#! - the ASSET is a fungible asset.
+#! - the ASSET is not associated with the faucet the transaction is being executed against.
+export.is_non_fungible_asset_issued
+    # assert that the asset is associated with the faucet the transaction is being executed against
+    # and that the asset is valid
+    exec.account::get_id swap drop exec.asset::validate_non_fungible_asset_origin
+    # => [ASSET]
+
+    # fetch the root of the SMT containing the non-fungible assets
+    exec.account::get_faucet_storage_data_slot exec.account::get_item
+    # => [SMT_ROOT, ASSET]
+
+    # prepare stack for obtaining the non-fungible asset from the tracking SMT
+    swapw exec.asset_vault::build_non_fungible_asset_vault_key
+    # => [ASSET_KEY, SMT_ROOT]
+
+    # get the non-fungible asset from the tracking SMT
+    exec.smt::get
+    # => [ASSET, SMT_ROOT]
+
+    # Assert the `ASSET` is an EMPTY_WORD, indicating that the non-fungible asset has not been 
+    # issued yet. We only need to check ASSET[3] as this is always set to the faucet_id_prefix
+    # and can not be 0 (in reversed stack order it will be top stack element).
+    swapw dropw neq.0 movdn.3 drop drop drop  
```

Comment:
We're not `assert`ing here, so this comment is slightly incorrect. Maybe change it to say just that we're checking whether the nf asset is issued or not.

Reply:
Ah, indeed, I forgot to update this comment. Thank you!